File: iosApp/ESMira/iosNative/statistics/ChartViewWrapper.swift
//
// Created by JodliDev on 06.10.20.
//

import Foundation
import Charts
import sharedCode
import UIKit

class ChartViewWrapper : ChartViewInterface {
	class AxisFormatter: IAxisValueFormatter {
		let statisticFormatter: ChartFormatterInterface
		
		init(_ statisticFormatter: ChartFormatterInterface) {
			self.statisticFormatter = statisticFormatter
		}
		
		func stringForValue(_ value: Double, axis: AxisBase?) -> String {
			return statisticFormatter.getString(value: Float(value))
		}
	}
	
	class IOsLegend : ChartViewInterfaceLegend {
		let legend: Legend
		
		init(_ legend: Legend) {
			self.legend = legend
		}
		
		
		func setVerticalAlignment(x: ChartViewInterfaceLegendVerticalAlignment) {
			switch(x) {
				case .top:
					legend.verticalAlignment = Legend.VerticalAlignment.top
				case .center:
					legend.verticalAlignment = Legend.VerticalAlignment.center
				case .bottom:
					legend.verticalAlignment = Legend.VerticalAlignment.bottom
				default:
					print("setPosition not implemented")
					return
			}
		}
		func setHorizontalAlignment(x: ChartViewInterfaceLegendHorizontalAlignment) {
			switch(x) {
				case .left:
					legend.horizontalAlignment = Legend.HorizontalAlignment.left
				case .center:
					legend.horizontalAlignment = Legend.HorizontalAlignment.center
				case .right:
					legend.horizontalAlignment = Legend.HorizontalAlignment.right
				default:
					print("setPosition not implemented")
					return
			}
		}
		
		func setOrientation(x: ChartViewInterfaceLegendOrientation) {
			switch(x) {
				case .vertical:
					legend.orientation = Legend.Orientation.vertical
				case .horizontal:
					legend.orientation = Legend.Orientation.horizontal
				default:
					print("setPosition not implemented")
					return
			}
		}
		
		
	}
	class IOsAxis : ChartViewInterfaceAxis {
		let axis: AxisBase
		
		init(_ axis: AxisBase) {
			self.axis = axis
		}
		func setAxisMaximum(x: Float) {
			axis.axisMaximum = Double(x)
		}
		
		func setAxisMinimum(x: Float) {
			axis.axisMinimum = Double(x)
		}
		
		func setDrawAxisLine(x: Bool) {
			axis.drawAxisLineEnabled = x
		}
		
		func setDrawGridLines(x: Bool) {
			axis.drawGridLinesEnabled = x
		}
		
		func setDrawLabels(x: Bool) {
			axis.drawLabelsEnabled = x
		}
		
		func setLabelCount(x: Int32) {
			axis.labelCount = Int(x)
		}
		
		func setLabelRotationAngle(x: Float) {
			(axis as! XAxis).labelRotationAngle = CGFloat(x)
		}
		
		func setPosition(x: ChartViewInterfaceAxisPosition) {
			switch(x) {
				case .top:
					(axis as! XAxis).labelPosition = .top
				case .topinside:
					(axis as! XAxis).labelPosition = .topInside
				case .bottom:
					(axis as! XAxis).labelPosition = .bottom
				case .bottominside:
					(axis as! XAxis).labelPosition = .bottomInside
				case .bothsided:
					(axis as! XAxis).labelPosition = .bothSided
				default:
					print("setPosition not implemented")
					return
			}
			
		}
		
		func setSpaceMax(x: Float) {
			axis.spaceMax = Double(x)
		}
		
		func setSpaceMin(x: Float) {
			axis.spaceMin = Double(x)
		}
		
		func setValueFormatter(x: ChartFormatterInterface) {
			axis.valueFormatter = AxisFormatter(x)
		}
		
		func setGranularity(x: Float) {
			axis.granularity = Double(x)
		}
		
	}
	
	let chartView: ChartViewBase
	init(_ chartView: ChartViewBase) {
		self.chartView = chartView
	}
	
	func getDataSetCount() -> Int32 {
		Int32(chartView.data?.dataSetCount ?? 0)
	}
	
	func getLeftAxis() -> ChartViewInterfaceAxis {
		return IOsAxis((chartView as! BarLineChartViewBase).leftAxis)
	}
	
	func getLegend() -> ChartViewInterfaceLegend {
		return IOsLegend(chartView.legend)
	}
	
	func getRightAxis() -> ChartViewInterfaceAxis {
		return IOsAxis((chartView as! BarLineChartViewBase).rightAxis)
	}
	
	
	func getXAxis() -> ChartViewInterfaceAxis {
		return IOsAxis(chartView.xAxis)
	}
	
	
	func setDescriptionEnabled(x: Bool) {
		//missing
	}
	
	func setDoubleTapToZoomEnabled(x: Bool) {
		(chartView as! BarLineChartViewBase).doubleTapToZoomEnabled = x
	}
	
	func setHighlightPerDragEnabled(x: Bool) {
		(chartView as! BarLineChartViewBase).highlightPerDragEnabled = x
	}
	
	func setHighlightPerTapEnabled(x: Bool) {
		chartView.highlightPerTapEnabled = x
	}
	
	
	func setDrawHoleEnabled(x: Bool) {
		(chartView as! PieChartView).drawHoleEnabled = x
	}
	
	func setEntryLabelColor(color: String) {
		(chartView as! PieChartView).entryLabelColor = DataSetWrapper.getNSUIColor(color)
	}
	
	func setMinOffset(x: Float) {
		(chartView as! BarLineChartViewBase).minOffset = CGFloat(x)
	}
	
	func setMinimHeight(x: Int32) {
		//TODO
	}
	
	func setScaleYEnabled(x: Bool) {
		(chartView as! BarLineChartViewBase).scaleYEnabled = x
	}
	
	func setVisibleXRangeMinimum(x: Float) {
		(chartView as! BarLineChartViewBase).setVisibleXRangeMinimum(Double(x))
	}
	
	func notifyDataSetChanged() {
		chartView.notifyDataSetChanged()
	}
	func notifyCurrentDataChanged() {
		chartView.data?.notifyDataChanged()
	}
	func fitScreen() {
		(chartView as! BarLineChartViewBase).fitScreen()
	}
	func zoom(scaleX: Float, scaleY: Float, x: Float, y: Float) {
		(chartView as! BarLineChartViewBase).zoom(scaleX: CGFloat(scaleX), scaleY: CGFloat(scaleY), x: CGFloat(x), y: CGFloat(y))
	}
	func moveViewToX(x: Float) {
		(chartView as! BarLineChartViewBase).moveViewToX(Double(x))
	}
	
	func getView() -> Any {
		chartView
	}
	}

File: androidApp/src/main/java/at/jodlidev/esmira/androidNative/statistics/ChartViewWrapper.kt
package at.jodlidev.esmira.androidNative.statistics

import at.jodlidev.esmira.sharedCode.statistics.ChartViewInterface
import at.jodlidev.esmira.sharedCode.statistics.ChartFormatterInterface
import com.github.mikephil.charting.charts.BarLineChartBase
import com.github.mikephil.charting.charts.Chart
import com.github.mikephil.charting.charts.PieChart
import com.github.mikephil.charting.components.AxisBase
import com.github.mikephil.charting.components.Legend
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.formatter.IAxisValueFormatter

/**
 * Created by JodliDev on 06.10.2020.
 */
class ChartViewWrapper(val chartView: Chart<*>): ChartViewInterface {
	
	class AndroidFormatter(private val formatter: ChartFormatterInterface): IAxisValueFormatter {
		override fun getFormattedValue(value: Float, axis: AxisBase?): String {
			return formatter.getString(value)
		}
	}
	class AndroidLegend(private val legend: Legend): ChartViewInterface.Legend {
		override fun setVerticalAlignment(x: ChartViewInterface.Legend.VerticalAlignment) {
			legend.verticalAlignment = when(x) {
				ChartViewInterface.Legend.VerticalAlignment.Top -> Legend.LegendVerticalAlignment.TOP
				ChartViewInterface.Legend.VerticalAlignment.Center -> Legend.LegendVerticalAlignment.CENTER
				ChartViewInterface.Legend.VerticalAlignment.Bottom -> Legend.LegendVerticalAlignment.BOTTOM
			}
		}
		
		override fun setHorizontalAlignment(x: ChartViewInterface.Legend.HorizontalAlignment) {
			legend.horizontalAlignment = when(x) {
				ChartViewInterface.Legend.HorizontalAlignment.Left -> Legend.LegendHorizontalAlignment.LEFT
				ChartViewInterface.Legend.HorizontalAlignment.Center -> Legend.LegendHorizontalAlignment.CENTER
				ChartViewInterface.Legend.HorizontalAlignment.Right -> Legend.LegendHorizontalAlignment.RIGHT
			}
		}
		
		override fun setOrientation(x: ChartViewInterface.Legend.Orientation) {
			legend.orientation = when(x) {
				ChartViewInterface.Legend.Orientation.Vertical -> Legend.LegendOrientation.VERTICAL
				ChartViewInterface.Legend.Orientation.Horizontal -> Legend.LegendOrientation.HORIZONTAL
			}
		}
	}
	class AndroidAxis(private val axis: AxisBase): ChartViewInterface.Axis {
		override fun setAxisMaximum(x: Float) {
			axis.axisMaximum = x
		}
		
		override fun setAxisMinimum(x: Float) {
			axis.axisMinimum = x
		}
		
		override fun setDrawAxisLine(x: Boolean) = axis.setDrawAxisLine(x)
		
		
		override fun setDrawGridLines(x: Boolean) = axis.setDrawGridLines(x)
		
		
		override fun setDrawLabels(x: Boolean) = axis.setDrawLabels(x)
		
		
		override fun setLabelCount(x: Int) {
			axis.labelCount = x
		}
		
		override fun setLabelRotationAngle(x: Float) {
			(axis as XAxis).labelRotationAngle = x
		}
		
		override fun setPosition(x: ChartViewInterface.Axis.Position) {
			(axis as XAxis).position = when(x) {
				ChartViewInterface.Axis.Position.Top -> XAxis.XAxisPosition.TOP
				ChartViewInterface.Axis.Position.TopInside -> XAxis.XAxisPosition.TOP_INSIDE
				ChartViewInterface.Axis.Position.Bottom -> XAxis.XAxisPosition.BOTTOM
				ChartViewInterface.Axis.Position.BottomInside -> XAxis.XAxisPosition.BOTTOM_INSIDE
				ChartViewInterface.Axis.Position.BothSided -> XAxis.XAxisPosition.BOTH_SIDED
			}
		}
		
		override fun setSpaceMax(x: Float) {
			axis.spaceMax = x
		}
		
		override fun setSpaceMin(x: Float) {
			axis.spaceMin
		}
		
		override fun setValueFormatter(x: ChartFormatterInterface) {
			axis.valueFormatter = AndroidFormatter(x)
		}

		override fun setGranularity(x: Float) {
			axis.granularity = x
		}

	}
	
	override fun getDataSetCount(): Int {
		return chartView.data.dataSetCount
	}
	
	override fun getLegend(): ChartViewInterface.Legend {
		return AndroidLegend(chartView.legend)
	}
	
	override fun getXAxis(): ChartViewInterface.Axis {
		return AndroidAxis(chartView.xAxis)
	}
	override fun getLeftAxis(): ChartViewInterface.Axis {
		return AndroidAxis((chartView as BarLineChartBase<*>).axisLeft)
	}
	override fun getRightAxis(): ChartViewInterface.Axis {
		return AndroidAxis((chartView as BarLineChartBase<*>).axisRight)
	}
	
	
	
	override fun setDoubleTapToZoomEnabled(x: Boolean) {
		(chartView as BarLineChartBase<*>).isDoubleTapToZoomEnabled = x
	}
	
	override fun setDrawHoleEnabled(x: Boolean) {
		(chartView as PieChart).isDrawHoleEnabled = x
	}
	
	override fun setEntryLabelColor(color: String) {
		(chartView as PieChart).setEntryLabelColor(DataSetWrapper.getIntColor(color))
	}
	
	override fun setScaleYEnabled(x: Boolean) {
		(chartView as BarLineChartBase<*>).isScaleYEnabled = x
	}
	
	override fun setMinOffset(x: Float) {
		(chartView as BarLineChartBase<*>).minOffset = x
	}
	
	override fun setVisibleXRangeMinimum(x: Float) = (chartView as BarLineChartBase<*>).setVisibleXRangeMinimum(x)
	
	override fun setDescriptionEnabled(x: Boolean) {
		chartView.description.isEnabled = x
	}
	override fun setHighlightPerTapEnabled(x: Boolean) {
		chartView.isHighlightPerTapEnabled = x
	}
	override fun setHighlightPerDragEnabled(x: Boolean) {
		(chartView as BarLineChartBase<*>).isHighlightPerDragEnabled = x
	}
	
	
	override fun notifyDataSetChanged() {
		chartView.notifyDataSetChanged()
		chartView.invalidate()
	}
	
	override fun notifyCurrentDataChanged() {
		chartView.data.notifyDataChanged()
	}
	
	override fun fitScreen() = (chartView as BarLineChartBase<*>).fitScreen()
	override fun zoom(scaleX: Float, scaleY: Float, x: Float, y: Float) = (chartView as BarLineChartBase<*>).zoom(scaleX, scaleY, x, y)
	override fun moveViewToX(x: Float) {
		chartView.post {
			(chartView as BarLineChartBase<*>).moveViewToX(x)
			
		}
	}
	
	override fun getView() : BarLineChartBase<*> {
		return chartView as BarLineChartBase<*>
	}
}